### ISSUE TITLE
(Lazy) Consider to use LAZY Preemption as default on desktop kernels

### ISSUE THREAD

**Issue by ptr1337** [CachyOS/linux-cachyos#339]
Lazy Preemption has been recently added to the 6.13 Kernel.
https://lwn.net/Articles/994322/

Generally, Lazy should replace "None" and "Voluntary", while maintaining for RT tasks still the full preemption modell.
We could adjust games in the future to use real time ISO, which allows to use at these full preemption, while other tasks still benefits from using the higher throughput.

Production Benchmarks generally looking good with Lazy Preemption.
Gaming needs to be evaluated.

**Comment by 1Naim** [comment]
I have been tire~lessly~ trying to benchmark this. [Here](https://paste.cachyos.org/p/04a75bf.diff) are my latest benchmarks of full vs lazy with [unixbench](https://aur.archlinux.org/packages/unixbench). Apart from the consistent regression in FIO across numerous runs, it's basically a cointoss between the two preemption models. This has also been observed in some elapsed compilations I made between full vs lazy (PREEMPT_DYNAMIC enabled in both kernels)

Mike's [results](https://lore.kernel.org/all/579b7ea34ef6e2f7c955abdfc0929fe1af36faef.camel@gmx.de/) from LKML also show that there's not much difference if at all between full vs lazy. I notice how the results between preemption models get tighter when dynamic preemption is enabled, so maybe we should do these benchmarks with it disabled. On the bright side, lazy preemption shows good results with PREEMPT_RT so maybe we can have that enabled to regain the huge performance loss that comes with realtime preemption.

TL;DR we should probably just keep full. I'm going to redo the compilation tests with the kernels [here](https://share.cachyos.org/x86_64_v3/kernel/preempt/). They're both built with dynamic preemption disabled.

**Comment by 1Naim** [comment]
Rejoice lazy preemption lovers. The elapsed compile time benchmarks with dynamic preemption have converged and there is a performance boost (albeit small).

Compiling mesa:
full: 424.69 423.83 424.81 - avg:  424.44 (sec)
lazy: 422.51 422.73 422.72 - avg: 422.65 (sec)

Improvements might be larger in micro-benchmarks.

----------------------------
For reference on how dynamic preemption can skewer results, here are results from a month ago with 2 different kernels with dynamic preemption enabled
full: 439.92 442.13 444.56
lazy: 443.08 443.41 443.11

**Comment by 1Naim** [comment]
Lazy preemption fails quite hard in high load, see https://flightlesssomething.ambrosia.one/benchmark/950.

**Comment by 1Naim** [comment]
> For reference on how dynamic preemption can skewer results

Finally realized what happened here. RCUs aren't preemptible with dynamic preemption off. PREEMPT_RCU would always be disabled even if I try to explicitly enable it.

> Lazy preemption fails quite hard in high load, see https://flightlesssomething.ambrosia.one/benchmark/950.

What's happening here is also likely because of PREEMPT_RCU being disabled. Maybe PREEMPT_DYNAMIC isn't broken after all.

**Comment by fridrichgood** [comment]
> > For reference on how dynamic preemption can skewer results
> 
> Finally realized what happened here. RCUs aren't preemptible with dynamic preemption off. PREEMPT_RCU would always be disabled even if I try to explicitly enable it.
> 
> > Lazy preemption fails quite hard in high load, see https://flightlesssomething.ambrosia.one/benchmark/950.
> 
> What's happening here is also likely because of PREEMPT_RCU being disabled. Maybe PREEMPT_DYNAMIC isn't broken after all.

I started looking for information based on your observations and found a series of patches. In my free time, I will check if they are in the current kernel. I'm using a modified 6.13.3-zen now.

[https://lwn.net/ml/all/20241213040658.2282640-1-ankur.a.arora@oracle.com/](url)

**Comment by 1Naim** [comment]
Those patches are submitted for 6.15. If you can provide some of your observations too regarding full vs lazy, that would be nice :)

**Comment by fridrichgood** [comment]
Ok, i used these patches and compile 6.13.5-zen1. So far, I haven't noticed the difference between full and lazy preempt. I used passmark and unigine superposition benchmark. Couldn't record mangohud results in unigine, but the difference isn't noticeable. Got the same results. Needs to be tested in other scenarios and intense games.

[passmark.txt](https://github.com/user-attachments/files/19109631/passmark.txt)

**Comment by 1Naim** [comment]
You probably won't notice any differences in gaming. Some other benchmarks or tasks that are more CPU-intensive may be more helpful.

**Comment by fridrichgood** [comment]
> You probably won't notice any differences in gaming. Some other benchmarks or tasks that are more CPU-intensive may be more helpful.

I guess I need parallel heavy tasks like compiling and watching 4k videos and other things at the same time. Hmm, but how can I catch these metrics?

**Comment by 1Naim** [comment]
You can use `time` to measure the elapsed time of a compilation.

**Comment by fridrichgood** [comment]
> You can use `time` to measure the elapsed time of a compilation.

Same results.

full
real	37m32,661s
user	356m17,633s
sys	23m3,506s

lazy
real	38m39,446s
user	355m59,014s
sys	22m58,798s


Also i tried tools from rt-tests package and i did not get noticeable results. Reference article:

https://www.opensourceforu.com/2021/05/an-analysis-of-the-real-time-performance-of-linux-kernels/

**Comment by oech3** [comment]
Link to ticket in upstream: https://gitlab.archlinux.org/archlinux/packaging/packages/linux/-/issues/123

I just noted.

**Comment by ptr1337** [comment]
Lets recheck due upstream kernel linux intend to have "this as default"

https://www.phoronix.com/news/Linux-Restrict-Preempt-Modes

The patch needs to be applied on top: https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/patch/?id=7dadeaa6e851e7d67733f3e24fc53ee107781d0f

**Comment by pallaswept** [comment]
The comments on the above phoronix article mentioned this issue, with the context being performance issues that had been noted with lazy preempt. I clicked that link as I'd noted the same.

A couple of suggestions I might offer for benchmarking this: 

Using [powersave](https://github.com/CachyOS/linux-cachyos/issues/339#issuecomment-2574910383) mode might not be the best choice, as the EPP preferences for this (as well as other system changes for example limits set by `tuned`) put a serious choke on the CPU. I think it's fair to say it's appropriate to test this using balanced (to reflect realistic usage) or performance (to remove any bottlenecks this is most correct) EPP preference.

`rtla timerlat` is useful to ensure that IRQs (and timers in general) are handled in a timely fashion - lazy preempt performed very badly for me under high load with low-latency audio, as reported in the phoronix comments that brought me here. It also performed quite badly with respect to frametime variance in games. Both were results of the timing of the respective drivers becoming very 'loose'. This tool will provide metrics that reflect that behaviour.

I hope these are helpful hints. Keep up the great work, Cachy nerds.

**Comment by 1Naim** [comment]
> Using [powersave](https://github.com/CachyOS/linux-cachyos/issues/339#issuecomment-2574910383) mode might not be the best choice, as the EPP preferences for this (as well as other system changes for example limits set by `tuned`) put a serious choke on the CPU. I think it's fair to say it's appropriate to test this using balanced (to reflect realistic usage) or performance (to remove any bottlenecks this is most correct) EPP preference

That was the powersave governor. EPP used there was balance_performance.

**Comment by 1Naim** [comment]
I think another thing to consider is disabling PREEMPT_DYNAMIC to inch out a bit more performance, since enabling it introduces runtime overhead. I don't think having it disabled since a major issue for us, since we provide various kernel flavours as supposed to Arch for instance where there is only one kernel so having PREEMPT_DYNAMIC is favourable there.

**Comment by pallaswept** [comment]
> > Using [powersave](https://github.com/CachyOS/linux-cachyos/issues/339#issuecomment-2574910383) mode might not be the best choice, as the EPP preferences for this (as well as other system changes for example limits set by `tuned`) put a serious choke on the CPU. I think it's fair to say it's appropriate to test this using balanced (to reflect realistic usage) or performance (to remove any bottlenecks this is most correct) EPP preference
> 
> That was the powersave governor. EPP used there was balance_performance.

"Performance" would be both the governor and EPP preference of choice for benchmarking where bottlenecks should be removed. Of course, all that gets complex because in the case of the governor, the reduced clock speeds on powersave, reduce temps, which allow higher boost clocks, and higher overall efficiency. That implies a need for  heatsoaking the CPU before the benchmark...

Anyway, I don't want to detract from the primary point of my reply which was: Test for jitter, because those of us who have, found that it was a problem. If you ignore the other advice about removing bottlenecks it'll only serve to exacerbate that problem, so maybe powersave wouldn't be such a bad thing :D

**Comment by sheviks** [comment]
Since we are discussing `PREEMPT_DYNAMIC` and `lazy` preemption here, I'd like to bring up something regarding the `rt` builds. 

For newer kernels (like 6.19 and 7.0+), we might be able to remove the following `if` condition in the `PKGBUILD`:

```bash
    # We should not set up the PREEMPT for RT kernels
    if [[ "$_cpusched" != "rt" && "$_cpusched" != "rt-bore" ]]; then
```

Currently, `PREEMPT_RT` mainly affects the underlying spinlock mechanism, whereas `PREEMPT_DYNAMIC` (full / lazy) governs the scheduler policy.

These two are no longer mutually exclusive and can actually coexist. For example, in the 7.0 kernel, we can achieve four combinations scaling from throughput-focused to latency-focused:
generic + lazy -> generic + full -> rt + lazy -> rt + full

Given this, we could probably drop this if restriction so that RT builds can also properly benefit from dynamic preemption configurations.

Thanks!

**Comment by akkiirah** [comment]
Please consider enabling `CONFIG_PREEMPT_DYNAMIC=y` again for normal/bore/bore-lto kernels, while keeping Full as the default.

With `linux-cachyos-bore-lto 7.0.3-1` I currently get:

```
$ zcat /proc/config.gz | grep CONFIG_PREEMPT_DYNAMIC
CONFIG_PREEMPT_DYNAMIC is not set
```

This means Full is effectively hardcoded and `preempt=lazy` no longer works.

For my mixed desktop/workstation use case, gaming, game development, Blender, and music production, Lazy performs much better in graphics-heavy workloads. In Blender and Cyberpunk 2077 I see severe throughput regressions with Full, in some cases roughly down to 1/4 of the FPS I had with Lazy on linux-cachyos-bore-lto 6.19.x, including severe artifacts and lag spikes.

Therefore I’d really like to get the runtime choice back. #806 already mentions that Lazy can bring back throughput quite a bit, which is exactly what is missing for my workload now that Dynamic is disabled.

**Comment by pallaswept** [comment]
In the metrics I just collected, lazy got me about 3% increase in throughput (measured with tbench per the lkml thread linked in the thread linked above) and about 300% increase in jitter (measured with rtla as above) when compared to full.

This obviously wants more testing. I know that isn't a popular opinion but I'm trying to see cachy be even better, not me be more popular.

**Comment by akkiirah** [comment]
The point is to give users the option again.

If Full remains the default, that is fine. But with `CONFIG_PREEMPT_DYNAMIC` disabled, `preempt=lazy` no longer works, so users who need Lazy for specific workloads have to rebuild the kernel.

Please re-enable `CONFIG_PREEMPT_DYNAMIC=y` for the normal/bore/bore-lto kernels, while keeping Full as the default.

**Comment by ptr1337** [comment]
> The point is to give users the option again.
> 
> If Full remains the default, that is fine. But with `CONFIG_PREEMPT_DYNAMIC` disabled, `preempt=lazy` no longer works, so users who need Lazy for specific workloads have to rebuild the kernel.
> 
> Please re-enable `CONFIG_PREEMPT_DYNAMIC=y` for the normal/bore/bore-lto kernels, while keeping Full as the default.

We already enabled it ,see last commit.
it will come with 7.0.4.

**Comment by akkiirah** [comment]
Thanks, much appreciated.

Sorry for the noise. I missed the commit and assumed nothing had changed yet because there was no reply in the issue.

Looking forward to 7.0.4.
